nomad/core_sched.go

```diff
@@ -706,9 +715,116 @@ func allocGCEligible(a *structs.Allocation, job *structs.Job, gcTime time.Time,
 	return timeDiff > interval.Nanoseconds()
 }
 
-// csiVolumeGC is used to garbage collect CSI volume publications
-func (c *CoreScheduler) csiVolumePublicationGC(eval *structs.Evaluation) error {
-	// TODO: implement me!
-	c.logger.Trace("garbage collecting unclaimed CSI volume publications")
+// csiVolumeClaimGC is used to garbage collect CSI volume claims
+func (c *CoreScheduler) csiVolumeClaimGC(eval *structs.Evaluation) error {
+	c.logger.Trace("garbage collecting unclaimed CSI volume claims")
+
+	// JobID smuggled in with the eval's own JobID
+	var jobID string
+	evalJobID := strings.Split(eval.JobID, ":")
+	if len(evalJobID) != 2 {
+		c.logger.Error("volume gc called without jobID")
+		return nil
+	}
+
+	jobID = evalJobID[1]
+	job, err := c.srv.State().JobByID(nil, eval.Namespace, jobID)
+	if err != nil || job == nil {
+		c.logger.Trace(
+			"cannot find job to perform volume claim GC. it may have been garbage collected",
+			"job", jobID)
+		return nil
+	}
+	c.volumeClaimReap([]*structs.Job{job}, eval.LeaderACL)
 	return nil
 }
+
+// volumeClaimReap contacts the leader and releases volume claims from
+// terminal allocs
+func (c *CoreScheduler) volumeClaimReap(jobs []*structs.Job, leaderACL string) error {
+	ws := memdb.NewWatchSet()
+	var result *multierror.Error
+
+	for _, job := range jobs {
+		c.logger.Trace("garbage collecting unclaimed CSI volume claims for job", "job", job.ID)
+		for _, taskGroup := range job.TaskGroups {
+			for _, tgVolume := range taskGroup.Volumes {
+				if tgVolume.Type != structs.VolumeTypeCSI {
+					continue // filter to just CSI volumes
+				}
+				volID := tgVolume.Source
+				vol, err := c.srv.State().CSIVolumeByID(ws, volID)
+				if err != nil {
+					result = multierror.Append(result, err)
+					continue
+				}
+				if vol == nil {
+					c.logger.Trace("cannot find volume to be GC'd. it may have been deregistered",
+						"volume", volID)
+					continue
+				}
+				vol, err = c.srv.State().CSIVolumeDenormalize(ws, vol)
+				if err != nil {
+					result = multierror.Append(result, err)
+					continue
+				}
+
+				gcAllocs := []*structs.Allocation{}
+				unclaimedNodes := map[string]struct{}{}
+
+				collectFunc := func(allocs map[string]*structs.Allocation) {
+					for _, alloc := range allocs {
+						// we call denormalize on the volume above to populate
+						// Allocation pointers. But the alloc might have been
+						// garbage collected concurrently, so if the alloc is
+						// still nil we can safely skip it.
+						if alloc == nil {
+							continue
+						}
+						if !alloc.Terminated() {
```

Comment:
Is this condition reversed? It looks like we're putting nodes in unclaimed if they have a running alloc and then calling controllerUnpublish on all the unclaimed nodes.

Reply:
Bah, I changed the name of the var to `unclaimedNodes` to avoid a double-negative elsewhere and reversed something in the process of that refactor. The test doesn't hit the controllerUnpublish branch so I missed it. Will fix.

Reply:
Fixed. Changed it up a bit by creating a slice of "known" nodes and then making a list of claimed nodes to test (because golang makes me sad by having no sensible set type).